### DISCUSSION TITLE
Nextjs styles from a shared library are not loaded on initial loading

### DISCUSSION THREAD

**Discussion by foxter111** [vercel/next.js#44420] [Help] [answered]
Hi everyone. I have my common UI library for use in different apps. I connect it via npm in my app, but I have a problem with the initial loading in NextJs. When the page loads I see the components blinking because the styles are not being applied to the components on the server. In my Ui library, I use styled-component for styling.

**NextJs version 12.2.2**

**For example**
This is how I see the form first second after load.
![Screenshot_5](https://user-images.githubusercontent.com/27918824/209937719-cf758cfc-b18c-46be-92fc-f3494a46a552.png)
This is how I see the form after load styles.
![Screenshot_6](https://user-images.githubusercontent.com/27918824/209937875-b78e2eab-5923-4191-83b9-54202860cbe0.png)

How can I solve this problem? I think, that I must add something in next.config or maybe in my _document.tsx page.

**Comment by icyJoseph** [comment, marked as answer]
Hi,

You need to make sure that Styled-Components walks your UI during SSR so that it can generate the first frame style tags and class names, in order to avoid flashing of un-styled content, during hydration.

Make sure your `_document` page looks like: https://github.com/vercel/next.js/blob/canary/examples/with-styled-components/pages/_document.tsx

You also need to tell the compiler that you are using styled-components, to make class names deterministic: https://github.com/vercel/next.js/blob/canary/examples/with-styled-components/next.config.js

## Solution

Reposting for visibility:

> Hi! no worries, but I haven't been able to try it out. I do image that the problem is how you package your styled components library. I think you need to set dependency to styled components as external - If you wanna take it further, do check how other libraries package their components, specifically for SSR.
> 
> For example Grommet, https://github.com/grommet/grommet/blob/master/webpack.config.babel.js - Notice that styled-components is labelled as external too.

OP had built their own component library. Setting `styled-components` as an external dependency was necessary to make it work during SSR.

**Comment by foxter111** [comment]
@icyJoseph 

I have this setting in these files.
It's my next.config
![Screenshot_7](https://user-images.githubusercontent.com/27918824/209944672-c478d045-9966-4d1c-b084-1ac4a66bcad2.png)
It's a part of my _document.tsx
![Screenshot_8](https://user-images.githubusercontent.com/27918824/209944733-78c58c46-0eac-4c2a-b611-259783e7fad5.png)

I use in Nextjs app styled-component too, it works nice, and every component gets css on the server. But the app doesn't see styles from my ui library, which uses styled-component too. 

I found a similar question here https://stackoverflow.com/questions/71260223/rollupd-react-component-css-styles-arent-being-added-to-ssr-document-in-nextjs , but it doesn't have an answer

**Reply by icyJoseph** [reply]
Ah, the plot thickens... It feels like the problem has to be within your library. There's many frameworks out there, which one can use without problem. Perhaps different library versions are used? Or some kind of ESM vs CJS conflict? Do you have a link to either of these repositories?

**Reply by foxter111** [reply]
Yes, I think, that I can share ui library with config https://github.com/foxter111/components-library .

**Reply by foxter111** [reply]
@icyJoseph I created the public package https://github.com/foxter111/components-library/pkgs/npm/components-library for the check problem. You can install this package in your new Nextjs app and you will see a problem.

**Reply by foxter111** [reply]
@icyJoseph Hi, I'm sorry, have you looked at this issue yet? I still don't know what's the reason...

**Reply by icyJoseph** [reply]
Hi! no worries, but I haven't been able to try it out. I do image that the problem is how you package your styled components library. I think you need to set dependency to styled components as external - If you wanna take it further, do check how other libraries package their components, specifically for SSR. 

For example Grommet, https://github.com/grommet/grommet/blob/master/webpack.config.babel.js - Notice that styled-components is labelled as external too.

**Reply by foxter111** [reply]
Oh, thanks a lot. You really helped me!! After I added styled components in external, my app began to render components on the server correctly.

**Reply by icyJoseph** [reply]
oh wow haha! Nice!!! I'll update the my original answer. Please don't forget to mark as answered!

**Reply by foxter111** [reply]
Yes, thanks again )
